src/main/java/bio/terra/app/controller/SearchApiController.java

```diff
@@ -94,7 +98,23 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
         @Valid @RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
         @Valid @RequestParam(value = "limit", required = false, defaultValue = "1000") Integer limit
     ) {
-        SearchQueryResultModel searchQueryResultModel = new SearchQueryResultModel();
+        List<String> accessibleIds =
+                iamService.listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT)
+                        .stream()
+                        .map(UUID::toString)
+                        .collect(Collectors.toList());
+
+        List<String> requestIds = searchQueryRequest.getSnapshotIds();
+
+        for (String snapShotId : requestIds) {
+            if (!accessibleIds.contains(snapShotId)){
+                throw new IamForbiddenException("User '" + getAuthenticatedInfo().getEmail()
+                        + "' does not have required action: " + IamAction.READ_DATA + " on snapshot id" + snapShotId);
+            }
+        }
+        List<String> idsToQuery = requestIds.isEmpty() ? accessibleIds : requestIds;
+
+        SearchQueryResultModel searchQueryResultModel = searchService.querySnapshot(searchQueryRequest, idsToQuery);
         return new ResponseEntity<>(searchQueryResultModel, HttpStatus.OK);
```

Comment:
it's totally minor but I find the `ResponseEntity` factory methods a bit easier to use. (It was a much bigger deal before java added the diamond operator)
```suggestion
        return ResponseEntity.ok(searchQueryResultModel);
```